### PR TITLE
Added average encoding Elbo objective

### PR DESCRIPTION
**Pull request by babak0032**
I added a different version of the Elbo loss (doc strings are added), which was based on:
http://approximateinference.org/accepted/HoffmanJohnson2016.pdf 

There are several ways that this could have been implemented (I'm open to all of them as the change should not be that hard), but I think computing average encoding values can be a trace method, and everything else can simply be in objective.

### CODE REVIEW

File: probtorch/stochastic.py [hunk 1]
```diff
@@ -321,6 +324,30 @@ def log_joint(self, sample_dim=None, batch_dim=None, nodes=None):
...
+
+        Arguments:
+            nodes: The subset of nodes for which the log pair probabilities to be computed.
+            sample_dim(int): The dimension that enumerates samples.
+            batch_dim(int): The dimension that enumerates batch items.
+        """
+
+        log_avg_joint = 0.0
+        log_avg_sep = 0.0
+        # Note: not entirely sure if this is correct for more than one node
```
Comment (jwvdm) [inline probtorch/stochastic.py]: Lol. This is the code review equivalent of writing "If you read this, then I will give you a bottle of Champagne" in your thesis.

File: probtorch/stochastic.py [hunk 2]
```diff
@@ -321,6 +324,30 @@ def log_joint(self, sample_dim=None, batch_dim=None, nodes=None):
...
+            if n in self._nodes:
+                node = self._nodes[n]
+                value = node.value
+                size = value.size()
+                values_expand = value.unsqueeze(3).repeat(1, 1, 1, size[batch_dim]).permute(batch_dim, sample_dim, 3, 2)
+                log_pair_probs = node.dist.log_prob(values_expand).permute(1, 0, 2, 3) # (SxBxBxD) dimension
+                log_avg_joint = log_avg_joint + log_mean_exp(log_pair_probs.sum(-1), 2)
+                log_avg_sep = log_avg_sep + log_mean_exp(log_pair_probs, 2).sum(-1)
+
+        return log_avg_joint, log_avg_sep
```
Comment (jwvdm) [inline probtorch/stochastic.py]: What is the variable `log_avg_sep` supposed to be here?

File: probtorch/stochastic.py [hunk 3]
```diff
@@ -321,6 +324,30 @@ def log_joint(self, sample_dim=None, batch_dim=None, nodes=None):
...
+        # Note: not entirely sure if this is correct for more than one node
+        for n in nodes:
+            if n in self._nodes:
+                node = self._nodes[n]
+                value = node.value
+                size = value.size()
+                values_expand = value.unsqueeze(3).repeat(1, 1, 1, size[batch_dim]).permute(batch_dim, sample_dim, 3, 2)
+                log_pair_probs = node.dist.log_prob(values_expand).permute(1, 0, 2, 3) # (SxBxBxD) dimension
+                log_avg_joint = log_avg_joint + log_mean_exp(log_pair_probs.sum(-1), 2)
+                log_avg_sep = log_avg_sep + log_mean_exp(log_pair_probs, 2).sum(-1)
```
Comment (jwvdm) [inline probtorch/stochastic.py]: These lines need some more comments – I'm not sure I understand what is going on here.

**Comment by alicanb** [comment]
Can we close this @babak0032 ?

**Pull request closed without merging**